Installation section, AcraStruct, Trying Docker

Lagovas opened the pull request:
Migrated `Getting started` + `Quickstart` sections from old doc to new as `Installation` with several pages that are standalone and was easy to migrate: `Trying Acra with docker` and `Digital Ocean marketplace`, plus migrated AcraStruct description to AcraBlocks that was mentioned in Installation.

I didn't migrate `Write an app with AcraWriter` from `Quickstart`. I think it should be migrated into other sections or rewritten.

storojs72 reviewed (commented):
Apart from minor typos, there are some common remarks for `acra/acra-in-depth/data-structures/_index.md`:

1) I would use common names of subsections for AcraBlock/AcraStruct sections: 
- Common structure
- Generation/Encryption
- Decryption

(we have now _Container structure / Generating / Decrypting_ for AcraBlocks and 
_Understanding AcraStruct / Generating AcraStruct / Decrypting AcraStruct_ for AcraStruct)

2) I would revisit Storage Models - since it is mostly oriented on AcraStruct. Is it the same for AcraBlocks? If not, that I would state explicitly in the beginning of this section.

3) I suppose that AcraStruct validator doesn't support AcraBlocks, so would state explicitly on it in the beginning of the section

4) I would add some details (as a separate section in the beginning) about pros and cons for AcraBlocks vs AcraStructs. What to select and for what cases. 

Other files seems to have appropriate structure

Lagovas commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,185 @@
+
+The storage model modes used in Acra are WholeCell and InjectedCell.
+
+In WholeCell mode, [AcraStruct](/pages/documentation-acra/#acrastruct) represents a complete piece of data (i.e. database cell, a file, or some data transmitted into [AcraTranslator](/pages/acratranslator/)). In this mode it is expected that the encrypted data will look something like:
```
AcraServer can work with both variants of AcraStruct storage. When it stored inside other container or as whole piece of data. AcraTranslator works only with the whole AcraStruct.

storojs72 replied:
Should we may be explicitly state on this detail? For example, we can write:
```
There are two storage model modes used in AcraServer: WholeCell and InjectedCell (AcraTranslator supports only WholeCell mode)
```

and this would allow to avoid explaining this details in further text of this section. WDYT?

Lagovas commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,185 @@
+
+**4.** Go to the [Acrustruct validator tab in the Documentation server header](/simulator/acra/):
+
+![](https://docs.cossacklabs.com/files/wiki/AcrastructValidator-link.png)
```
nice catch, I will try to find the correct link to image. Because there I had migrated content as is. So this image should not be rendered on our current docserver too...

Lagovas commented:
I had applied suggestions related to typos and rephrasing. And I will look one more time on the data structures description.

Lagovas commented:
@storojs72 , I renamed sections and now they are similar for AcraStruct and AcraBlock, plus I added examples of output for each type of struct. Can you review examples, is it useful in such form? Because I don't know how better to show examples of AcraStruct/Acrablock. Maybe you can suggest other way to describe?
Other your suggestions I applied too, so please, mark as resolved that are good for you now.

ilammy commented on acra/guides/trying-acra-with-docker/_index.md:
```diff
@@ -0,0 +1,403 @@
+bookCollapseSection: true
+---
+
+# Acra + Docker
+
+You can use containers with Acra. Acra works well with [Docker](https://www.docker.com/what-docker). However, we advise only going the "Acra in Docker containers" route for testing purposes, and NOT in a real-world production environment.
+
+This is why this document is called "Trying Acra with Docker", not "Using Acra with Docker in a Production Setting".
```
This is also a fairly large document. I believe it will be more readable if split into multiple pages.

ilammy commented on acra/guides/trying-acra-with-docker/_index.md:
```diff
@@ -0,0 +1,403 @@
+
+The examples contain references to `acra-keymaker` and `acra-authmanager` containers inside. They are used for creation and distribution of the necessary keys. They were included for simplification of the test launch and should not be used in production schemes (where the keys should be generated manually and deployed to an appropriate host according to the security rules of your infrastructure).
+
+Check out the [docker/README](https://github.com/cossacklabs/acra/blob/master/docker/README.md) for examples and descriptions of compose files.
```
And the README points back to documentation...

**recursion**
　　_see:_ recursion

ilammy reviewed (commented):
Well, it might be okay to migrate the docs as is, but the layout and content could definitely use more love, whether in this PR directly or later.

#### Empty files with no placeholders

You add all these empty files, you should at least include some metadata in them, like a title:

- `acra/configuring-maintaining/general-configuration/acra-server.md `
- `acra/configuring-maintaining/general-configuration/acra-translator.md`
- `acra/security-controls/transport-security/acra-connector.md`


#### `files/wiki/*`

Is it really necessary to put them under "wiki" subdirectory?

I know it’s like that right now, but it’s a pretty sloppy way to organize files if you ask me.


#### `acra/configuring-maintaining/installing`

You have a clean slate here, why not make it right? IMO, these are three separate steps: first you install Acra, then configure it, then maintain your installation. There is no need to cram all of these under a shared section.

- `acra/installation`
- `acra/configuration`
- `acra/maintenance`

This structure looks more useful to me.

storojs72 commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,185 @@
+* Forms a container: packs together Begin_Tag, length of all other parts, DEK encryption backend identifier, KEK identifier, DEK encryption backend identifier, encrypted DEK and ciphertext.
+
+### Decrypting
+
```
I see grammar improved ('s' added to verbs), but what about adding some leading phrase, like "AcraServer performs following steps" and further we enumerate steps: 'validates ... ', 'extracts ... '

storojs72 commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,185 @@
+2. `<AcraSctruct>`,
+3. `<File containing AcraStruct alongside other data>`.
+
+The main difference between these modes lies in performance. In the WholeCell mode, AcraStructs are simply decrypted. In InjectedCell mode, [AcraServer](/pages/documentation-acra/#server-side-acraserver) needs to find AcraStructs inside some other data element first and then decrypt them, which will obviously take longer. The process of searching for the necessary piece of data takes place as the ...
```
In new version - 'which takes obviously take longer' - one 'take' is redundant

storojs72 commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,185 @@
+- Connects attributes together as described in the original formula.
+- Erases/fills the memory area containing the `Throwaway_Keypair` and original payload with zeros.
+
+We recommend you to check out AcraStruct [examples](https://github.com/cossacklabs/acra/tree/master/examples) and try the [⚙️Acra Engineering Demo⚙️](https://github.com/cossacklabs/acra-engineering-demo/#what-is-this).
```
Seems it requires a bit of clarification: "to check out the process of creating the AcraStruct" or smth like that. WDYT?

storojs72 commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,226 @@
+
+AcraBlock example for plaintext: `example`:
+
+`[34 34 34 34 141 0 0 0 0 0 0 0 0 119 199 0 76 0 0 1 1 64 12 0 0 0 16 0 0 0 32 0 0 0 74 57 33 205 212 208 201 173 5 49 141 175 138 74 129 21 72 124 140 56 23 187 147 90 79 15 192 32 164 252 144 234 120 9 212 13 105 61 48 221 98 14 188 170 171 97 212 137 231 11 99 218 252 110 120 212 32 72 195 54 0 1 1 64 12 0 0 0 16 0 0 0 7 0 0 0 72 81 94 231 179 35 95 35 89 97 196 52 15 142 126 144 136 26 86 192 ...
+
+* Begin_Tag[4] - `[34 34 34 34]`
+* Rest_AcraBlock_Length[8] - `[141 0 0 0 0 0 0 0]`
+* Key_Encryption_Backend_Identifier[1] - `[0]`
+* Key_Encryption_Key_ID[2] - `[119 199]`
+* Data_Encryption_Backend_Identifier[1] - `[0]`
+* Data_Encryption_Key_Length[2] - `[76 0]`
+* Encrypted_Data_Encryption_Key[*] - `[0 1 1 64 12 0 0 0 16 0 0 0 32 0 0 0 74 57 33 205 212 208 201 173 5 49 141 175 138 74 129 21]`
+* Encrypted_Data[*] - `[72 124 140 56 23 187 147 90 79 15 192 32 164 252 144 234 120 9 212 13 105 61 48 221 98 14 188 170 171 97 212 137 231 11 99 218 252 110 120 212 32 72 195 54 0 1 1 64 12 0 0 0 16 0 0 0 7 0 0 0 72 81 94 231 179 35 95 35 89 97 196 52 15 142 126 144 136 26 86 192 3 92 233 23 106 136 192 110 24 237 45 62 109 105 70]`
```
IMO: For me it looks normally (don't know about typical user). Often it is required to analyze the output of raw bytes, and I'm accustomed seeing that say DataLength is 8 bytes - [51, 0, 0, 0, 0, 0, 0, 0] and they are definitely can by found in overall AcraStruct stream of bytes

ilammy replied:
It just looks sloppy. Like Acra developer just copied and pasted debugger output here, without giving any thought of how it would look, how it would read, how non-Acra-developers are going to make any sense of it, and what is the overall goal of this documentation in the first place. This checks the box “we have N kilobytes of text that explains something about AcraStruct internals” but not every text is documentation, let alone *good* documentation.

`[72 124 140 56 23 187 147 90 79 15 192 32 164 252 144 234 120 9 212 13 105 61 48 221 98 14 188 170 171 97 212 137 231 11 99 218 252 110 120 212 32 72 195 54 0 1 1 64 12 0 0 0 16 0 0 0 7 0 0 0 72 81 94 231 179 35 95 35 89 97 196 52 15 142 126 144 136 26 86 192 3 92 233 23 106 136 192 110 24 237 45 62 109 105 70]` is no more informative than “95 bytes”.

storojs72 reviewed (approved):
LGTM. I like changes introduced.

Minor: I resolved most of threads, but some are still questionable. Though it doesn't have too much of impact on overall quality

Lagovas commented:
@ilammy ,

> 1. I would use common names of subsections for AcraBlock/AcraStruct sections
Updated

> Is it really necessary to put them under "wiki" subdirectory?
> I know it’s like that right now, but it’s a pretty sloppy way to organize files if you ask me.

Agree, moved to `/files/acrastruct/*`

> You have a clean slate here, why not make it right? IMO, these are three separate steps: first you install Acra, then configure it, 

then maintain your installation. There is no need to cram all of these under a shared section.

For now, it's moved according to our new structure in wiki. It's not obvious how do it better now until we migrate all current docs and fill gaps with new content that may be large and push us to re-organize one more time (a lot of about TLS/vault/redis/keystore configuration/etc)


> I think this should be an ordered list, and the grammar can improved: imperative mood and active voice, “generate this”, 
“encrypt that”, etc. Ditto elsewhere.

I do not agree with ordered list because it's just explaining how it works in general. but some steps may be swapped (encrypt data first with DEK or encrypt DEK with KEK first). 

What about grammar... Actually, I applied @storojs72 suggestion ) And here we describe what Acra does from its view. I looked how Postgresql describes in same manner [1](https://www.postgresql.org/docs/10/ssl-tcp.html), [2](https://www.postgresql.org/docs/10/kernel-resources.html).

> This should definitely go to some page.

Agree, moved

> Also, iOS and Android are not languages. Mayonnaise is not a language either. And ORM is not really required.

updated to `Objective-C/Swift (iOS), or Java (Android).`

> Similarly, these should go to some separate “How to configure Acra after installation” page.

temporary moved to `acra/configuring-maintaining/general-configuration/_index.md` as a general description of what should be done before usage. Sure that we will extend this section with new suggestions to use TLS as default secure transport instead acra-connector.

> It's not very appropriate to put “quickstart” after the fold. This should be a separate page somewhere else. Or just a part of the introduction to installation on the index page.

Agree, removed header and left links to separate pages where described installation from github/repository/docker as next steps

> There's already a page for this...

Agree, removed and left link to guide `trying docker` in list with `installing from github/repository` options.

> This is also a fairly large document. I believe it will be more readable if split into multiple pages.

I don't agree. It's a guide where explained base structure of our `docker/` folder and docker-compose files and examples how to use it. IMHO, it's not so big page and looks good enough. Intro + structure + example. IMHO, it's not comfortable always jumping from page to page and open XX new tabs from content of one page just to read several paragraphs.

ilammy commented on acra/configuring-maintaining/general-configuration/_index.md:
```diff
@@ -2,3 +2,20 @@
+- Integrate AcraWriter closely following the [examples](https://github.com/cossacklabs/acra/tree/master/examples) and using either AcraConnector (if available on this machine) or direct TLS database connection.
+- Test everything:
+
+  **1.** Connect AcraConnector to AcraServer, send a regular request to your database through AcraConnector.<br/>_Result_: If you see the answer, AcraConnector and AcraWriter are able to connect and forward signals back and forth. It means that the network and the keys are fine.
+
+  **2.** Upon integrating AcraWriter into your code, try generating an [AcraStruct]({{< ref "acra/acra-in-depth/data-structures/#understanding-acrastruct" >}}) from some payload.<br/>_Result_: If you succeed in running AcraWriter code, Themis library is installed properly and some of the keys are located in the expected place.
+
+  **3.** Write a row with AcraStruct into the database, either directly or through AcraConnector. Request this row through AcraConnector.<br/>_Result_: If you see decrypted payload in the response, the scheme works properly.
```
I believe it would lead to overall better result if ordered lists are represented with ordered lists. You can use CSS to make numbers bold.

```markdown
  1. Connect AcraConnector to AcraServer, send a regular request to your database through AcraConnector.

     _Result_: If you see the answer, AcraConnector and AcraWriter are able to connect and forward signals back and forth.
     It means that the network and the keys are fine.

  2. Upon integrating AcraWriter into your code, try generating an [AcraStruct] from some payload.

     _Result_: If you succeed in running AcraWriter code,
     Themis library is installed properly and some of the keys are located in the expected place.

  3. Write a row with AcraStruct into the database, either directly or through AcraConnector.
     Request this row through AcraConnector.

     _Result_: If you see decrypted payload in the response, the scheme works properly.

[AcraStruct]: {{< ref "acra/acra-in-depth/data-structures/#understanding-acrastruct" >}}
```

Here and in other places.

ilammy replied:
My point is that if you write things like this:

```markdown
  **1.** One
  **2.** Two
  **3.** Three
```
then it will not render into `<ol>`, as opposed to this:
```markdown
  1. One
  2. Two
  3. Three
```
The latter form should be preferred. If the number absolutely must be bold, `ol::marker` in CSS will take care of that, globally, across all documentation.

ilammy commented on acra/acra-in-depth/data-structures/acrastruct-validator.md:
```diff
@@ -0,0 +1,53 @@
+
+## AcraStruct validator
+
+To decrypt AcraStructs generated by AcraWriter in your application, you can also use our [AcraStruct validator](/simulator/acra/). It works as AcraServer / AcraTranslator and uses private decryption keys.
```
I believe that either *all* relative URL paths should use `{{< ref "/my/relative/path" >}}` syntax, or none of them. Including URLs of images.

ilammy replied:
Oh! Now I understand this magic. Thanks 👍

ilammy commented:
> moved to `/files/acrastruct/*`

I believe some of these files are also used by Themis documentation. Be careful, you might need to update more links elsewhere.

ilammy commented:
Also, I'm going to hit this point every day until you like it.

```markdown
# Title of the page

## Section

### Subsection

### Subsection

#### Subsubsection

## Section

## Section
```

That is

```
#
##
###
####
##
##
```

not

```
###
####
####
#####
```

and not

```
####
##
##
#
###
```

Properly nested header structure, not whatever arbitrary levels they had when pieces of text are getting cut into smaller pages.

ilammy commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,180 @@
+
+## AcraBlock
+
+### Container structure
```
This should be moved a bit lower, after “To generate AcraBlocks Acra-Server uses symmetric keys”, where structure description actually starts.

ilammy commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,180 @@
+
+### Container structure
+
+AcraBlock is a container format. Acra uses an envelope encryption strategy: plaintext data is encrypted using data encryption key (DEK), and then DEK is encrypted using key encryption key (KEK). AcraBlock uses AES-GCM for both encryption procedures.
```
```markdown
AcraBlock is a cryptographic container which is more compact than AcraStruct.
AcraBlock uses an envelope encryption strategy [...]
```

ilammy commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,180 @@
+
+AcraBlock supports key rotation: it's possible to rotate only KEK without re-encrypting the data (known as "key rotation without data re-encryption"), or re-encrypt data and rotate both keys ("key rotation with data re-encryption").
+
+To generate AcraBlocks Acra-Server uses symmetric keys generated for every ClientID/ZoneID.
```
It seems that `AcraServer` is more common spelling.

ilammy commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,180 @@
+* `Secure Cell in Seal Mode` - `0`
+
+
+AcraBlock example for plaintext: `example`:
+
+`[34 34 34 34 141 0 0 0 0 0 0 0 0 119 199 0 76 0 0 1 1 64 12 0 0 0 16 0 0 0 32 0 0 0 74 57 33 205 212 208 201 173 5 49 141 175 138 74 129 21 72 124 140 56 23 187 147 90 79 15 192 32 164 252 144 234 120 9 212 13 105 61 48 221 98 14 188 170 171 97 212 137 231 11 99 218 252 110 120 212 32 72 195 54 0 1 1 64 12 0 0 0 16 0 0 0 7 0 0 0 72 81 94 231 179 35 95 35 89 97 196 52 15 142 126 144 136 26 86 192 ...
```
I think this should be in some “AcraBlock example” section.

(Ditto for AcraStructs.)

vixentael commented:
> Also, I'm going to hit this point every day until you like it.

Ahaha, @ilammy pushes people on a bright side of history :D

ilammy commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,180 @@
+
+In InjectedCell mode, CryptoEnvelope is stored inside some piece of data, i.e. inside some file or in a database cell with a file inside, with CryptoEnvelope as a piece of that file, not the whole file. In this mode, the encrypted data will look something like this:
+
+1. `<Some CryptoEnvelope data, some other CryptoEnvelope data>`,
+2. `<CryptoEnvelope>`,
+3. `<File containing CryptoEnvelope alongside other data>`.
```
This is very confusing. Do all three parts contain CryptoEnvelope? Or just one? Is the order important? How is this different from three CryptoEnvelopes above? So many questions, so few answers...

The point you need to carry across is that in WholeCell mode encrypted data looks like this:

```
+-------------------------+
|                         |
| AcraStruct or AcraBlock |
|                         |
+-------------------------+
```

That is, exactly one AcraStruct which is the only and entire thing a data piece contains.

While InjectedCell may look like this:

```
+-------------------------+
|  Arbitrary prefix data  |
+-------------------------+
|                         |
| AcraStruct or AcraBlock |
|                         |
+-------------------------+
|      Some more data     |
| that Acra retains as is |
+-------------------------+
|    Another AcraStruct   |
+-------------------------+
|    More non-Acra data   |
+-------------------------+
```

Where Acra will find and decrypt all embedded AcraStructs while leaving everything else intact.

Some specific examples would also help, I believe. IDK, some Protobuf with three blobs where only one of them in actually encrypted.

ilammy commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,180 @@
+
+Zone Ids are identifiers of EC keys used for matching Zones to keys. The current format is 8 bytes "begin tag" + 16 symbols a-zA-Z. For example: `DDDDDDDDzxzXVyBBaNclkgPS` where `DDDDDDDD` is "begin tag" and `zxzXVyBBaNclkgPS` is unique key identifier. 
+
+## When use AcraBlocks and when AcraStructs
+
+If you need end-to-end encryption then you should choose AcraStructs and encrypt data on client side using public key. In all other cases you should prefer AcraBlocks that have less output size, fewer keys size and are more performant because use only symmetric key encryption. When AcraStructs rely on asymmetric plus symmetric encryption algorithms and are slower. AcraServer supports transparent  ...
```
IMO, this should be somewhere closer to the top since this is something people would want to read. Making a choice between AcraBlocks and AcraStructs is important for the users. Implementation details and internal structure is something a user does not care much about, that's not their problem.

ilammy commented on acra/acra-in-depth/data-structures/_index.md:
```diff
@@ -2,3 +2,226 @@
 title: Data structures
 bookCollapseSection: true
```
> There's way too much information for one page. I believe AcraBlock and AcraStruct description should at least go on their own pages

Bump.

A better page structure would be something like this:

* An introductory page. Tell that there are two container formats: AcraBlock and AcraStruct. What are high-level differences between them. When I would want to choose one or the other.
  * Describe AcraBlock here.
  * Describe AcraStruct here.

ilammy commented:
@vixentael, just to clarify my intentions: these are all drive-by comments, I'm in no way blocking the merge and do not want to do that. There are lots of things that I personally would improve in data structure description, grammar, etc. But for that it would be easier to just submit a pull request than to spam a comment for each paragraph describing how I would improve it.

It might be perfectly reasonable approach to quickly migrate the text in whatever awful form it is, and then improve it later. But 1) typically “later” means “never” because sure as hell there are more important things to do than polishing documentation that almost no one will read, 2) this PR is open for almost three weeks and still not merged, so I guess you all *do* want to receive more feedback and improve it before merge.

ilammy commented on acra/configuring-maintaining/installing/installing-from-repository.md:
```diff
@@ -0,0 +1,88 @@
+* `$OS` should be `debian` or `ubuntu`.
+* `$RELEASE` should be one of Debian or Ubuntu release names. You can determine this by running `lsb_release -cs`, if you have `lsb_release` installed.
+
+We currently build packages for the following OSs and RELEASE combinations:
+
+- *CentOS 7*,
+- *Debian "Jessie" (Debian 8)*,
+- *Debian "Stretch" (Debian 9)*,
+- *Ubuntu Xenial Xerus (Ubuntu 16.04)*,
+- *Ubuntu Artful Aardvark (Ubuntu 17.10)*,
+- *Ubuntu Bionic Beaver (Ubuntu 18.04)*.
```
Blatant lies. The current list is this:

```markdown
- `debian jessie` – Debian 8 (Jessie)
- `debian stretch` – Debian 9 (Stretch)
- `ubuntu xenial` – Ubuntu 16.04 LTS (Xenial Xerus)
- `ubuntu bionic` – Ubuntu 18.04 LTS (Bionic Beaver)
```

I believe CentOS 7 should not be here, in a section about installation on Debian.

Also, it would be nice to provide packages for Debian 10, soon-to-be-released Debian 11, and Ubuntu 20.04. It's 2021, after all.

Lagovas replied:
Updated to 
```
- *Debian "Wheezy" (Debian 7)*,
- *Debian "Jessie" (Debian 8)*,
- *Debian "Stretch" (Debian 9)*,
- *Debian "Buster" (Debian 10)*,
- *Trusty Precise Pangolin (Ubuntu 12.04)*.
- *Trusty Trusty Tahr (Ubuntu 14.04)*.
- *Ubuntu Xenial Xerus (Ubuntu 16.04)*,
- *Ubuntu Yakkety Yak (Ubuntu 16.10)*.
- *Ubuntu Zesty Zapus (Ubuntu 17.04)*.
- *Ubuntu Artful Aardvark (Ubuntu 17.10)*,
- *Ubuntu Bionic Beaver (Ubuntu 18.04)*.
- *Ubuntu Focal Fossa (Ubuntu 20.04)*.
```
Checked versions from https://pkgs-ce.cossacklabs.com/stable/debian/dists/ and https://pkgs-ce.cossacklabs.com/stable/ubuntu/dists/

ilammy replied:
`dists` is only a list of available distributions in the repository. A package may be available only for some of them. Specifically, `acra` package is currently available only for the ones I listed: Debian 8 & 9, Ubuntu 16.04 & 18.04.

Lagovas replied:
How you check it? You are right. I looked on artful and found in `Packages` package `acra`. But can't download .deb *(
And your list is correct.

ilammy commented on acra/guides/advanced-integrations/digital-ocean-marketplace.md:
```diff
@@ -0,0 +1,27 @@
+
+You can configure AcraServer yourself or try it with pre-configured settings. Acra 1-Click App has an interactive configuration script that prepares AcraServer for work as a Transparent encryption proxy with a PostgreSQL database.
+
+With DigitalOcean Acra 1-Click App, you get Acra Community Edition which is totally free. Acra is [also available as paid Acra Pro and Acra Enterprise versions](https://www.cossacklabs.com/acra/#pricing) with additional features and extended support.
```
Acra Pro has been discontinued, IIRC.

Lagovas replied:
nice catch

ilammy commented on acra/guides/trying-acra-with-docker/_index.md:
```diff
@@ -0,0 +1,403 @@
+
+#### Compose files
+
+We wanted you to be able to easily try the most useful schemes that we prepared as Docker Compose files in the `docker` subdirectory. The name of each Docker Compose file describes its components and their interconnections in a simple form. For example: `docker-compose.pgsql-nossl-server-ssession-proxy.yml` is a scheme with Postgresql DB, AcraServer, and AcraConnector, connected to AcraServer thr ...
```
> We wanted you to be able to easily try [...]

And we still want. This is not an apology letter, present tense is better, IMO.

Lagovas replied:
+1

ilammy commented on acra/guides/trying-acra-with-docker/_index.md:
```diff
@@ -0,0 +1,403 @@
+
+Check out the [docker folder](https://github.com/cossacklabs/acra/blob/master/docker) for examples of compose files.
+
+Most likely you would like to set some variables before launch.
```
> Most likely you would like

I'd say “you would *need*” if not setting these variables was not an option.

However, Compose files plug in some defaults to just work™, so I'd say here, “The example Docker Compose files could be customized with environment variables before launch”, or something like that, sounding like this is a completely optional step. After all, these Compose files are explicitly *not* intended to be used in production, so there is no real danger in letting the users to use the defaults.

Similarly, the lines below should probably switch “please” for “you may” because it's not required for the users to configure anything, so we don't want to implore them to do so.

Alternatively, if you don't want that, you should force the users to set all these variables instead. But I think it will be less convenient that way.

ilammy commented on acra/guides/trying-acra-with-docker/_index.md:
````diff
@@ -0,0 +1,403 @@
+
+Most likely you would like to set some variables before launch.
+
+Please set `ACRA_MASTER_KEY` environmental variable, or check [KeyManagement]({{< ref "acra/acra-in-depth/cryptography-and-key-management/#-INVALID" >}}) page for more details:
+```bash
+
+export ACRA_SERVER_MASTER_KEY=$(echo -n "My_Very_Long_Key_Phrase_ge_32_chars" | base64)
+export ACRA_CONNECTOR_MASTER_KEY=$(echo -n "My_Very_Long_Key_Phrase_ge_32_chars" | base64)
+```
````
“Master keys for AcraServer and AcraConnector can be configured separately. Please read the key management guide for more details.”

There are two master keys, these are different from `ACRA_MASTER_KEY` with non-Compose setup, and you cannot use `ACRA_MASTER_KEY` with it. You must set these variables if you want to use different keys.

Also, the empty line after ` ```bash ` should probably come before that line.

ilammy commented on acra/guides/trying-acra-with-docker/_index.md:
````diff
@@ -0,0 +1,403 @@
+export ACRA_HTTPAUTH_PASSWORD=<http_auth_password>
+```
+
+Now you can run `docker-compose`:
+```bash
+    docker-compose -f docker/<compose_file_name> up
+```
+
+And connect to ports described above.
````
You don't need to indent code lines if you use fences already.

The “ports described above” were described *really, really* above, somewhere in quickstart section. You should at least put a link to it here. It would be better to just duplicate the port assignments for Acra services here.

ilammy commented on acra/guides/trying-acra-with-docker/_index.md:
```diff
@@ -0,0 +1,403 @@
+
+## Example application
+
+In the following examples, we assume that the schemes are running and all the environment variables are set as described in the code block below. For your convenience, we recommend that you perform the following actions in two different shell windows:  run the Docker in one and the examples in another.
```
> we recommend that you perform the following actions in two different shell windows

I find this very awkward and these “(shell window #​2)” annotations confusing. We don't do much terminal juggling here. Can we just suggest the users to run `docker-compose` in a separate terminal to observe Acra logs better? This would imply that other commands should be tried in a different terminal.

ilammy commented on acra/guides/trying-acra-with-docker/_index.md:
```diff
@@ -0,0 +1,403 @@
+### Install the dependencies for the example application
+(shell window #2)
+
+> Important: You must start with installing Themis encryption library for Acra to work. Please follow either [these instructions]({{< ref "themis/installation/" >}}) or use a simple [script](https://pkgs.cossacklabs.com/scripts/libthemis_install.sh) to install Themis before proceeding.
```
Hugo has its own blocks which would look better here:

```markdown
{{< hint warning >}}
**Important:**
You need to have Themis Core library installed [...]
{{< /hint >}}
```

But hey. This is stupid. You provide all these Docker Compose files for running Acra, why there isn't a premade image for running examples with all necessary dependencies already installed? That would allow users to try Acra out even on Windows. They would not need to install any shady Themis libraries just to try things out, and the instructions would be just `docker exec -it acra-examples-python /bin/bash` to dive into a container that has everything ready to go.

ilammy commented on acra/guides/trying-acra-with-docker/_index.md:
````diff
@@ -0,0 +1,403 @@
+sudo chown $(whoami) -R docker/.acrakeys
+```
+
+```bash
+python examples/python/example_without_zone.py \
+--public_key=docker/.acrakeys/acra-writer/${ACRA_CLIENT_ID}_storage.pub \
+    --db_user=${POSTGRES_USER} \
+    --db_password=${POSTGRES_PASSWORD} \
+    --host=127.0.0.1 \
+    --port=5432 \
+    --print
+```
````
OCD: `--public_key` is not indented.

Lagovas commented on acra/configuring-maintaining/installing/installing-from-github.md:
````diff
@@ -0,0 +1,98 @@
+```
+sudo useradd -m acra-server
+sudo su acra-server
+cd ~/
+export GOPATH=`pwd`
````
explicit better than implicit. If someone familiar with golang then he knows how it works and can skip it. If some newbie reads and want to try then better we describe steps that will work in 100% cases in most of environments. Plus we should show path to binary below. If we don't set this variable then it will be `''` and command `$GOPATH/bin/acra-connector` will expanded to `""/bin/acra-connector`. And we need to explain that if you didn't set GOPATH then you should call binary from `$HOME/go/bin/executable`. If you have set then use `%OPATH/bin/executable` and bla bla.

ilammy replied:
I'd still rather say once:

> Please make sure your GOBIN is in PATH:
> 
> ```
> export PATH=$PATH:${GOBIN:-${GOPATH:-$HOME}/go/bin}
> ```

and then proceed to use the same binary names that you would get if you installed Acra via packages.

Newbies who have no clue will just recite the spell in their terminal. People who have some Go experience will know where `go get` puts stuff and how to launch it.

Lagovas commented on acra/configuring-maintaining/installing/installing-from-github.md:
````diff
@@ -0,0 +1,98 @@
+$GOPATH/bin/acra-server --db_host=127.0.0.1
+```
+
+If you see an error message `"master key is empty"`, it means that you haven't generated `ACRA_MASTER_KEY`, please return to the [Key Generation step]({{< ref "acra/configuring-maintaining/installing/installing-from-github.md" >}}).
````
changed to another link where will be described simple way to generate all kind of keys for examples/tests

Lagovas commented on acra/configuring-maintaining/installing/installing-from-github.md:
```diff
@@ -0,0 +1,98 @@
+
+If you see error message similar to "Configuration error: AcraConnector private key .acrakeys/someid doesn't exists", it means that you haven't generated keys or keys are placed in a wrong folder, please return to the [Key Generation step]({{< ref "acra/configuring-maintaining/installing/#build-the-key-generator-and-generate-the-keys" >}}).
+
+AcraConnector is now listening on the localhost port **9494**.
```
added. but imho, it's obvious... it's like `install VLC to view movies and try to open video file to be sure`)

Lagovas reviewed (commented):
please add comments till 15:00, then I will merge it to master and deploy to staging. Comments added after that will apply in next PRs

ilammy commented on acra/configuring-maintaining/installing/_index.md:
```diff
@@ -2,3 +2,33 @@
+- Your application talking to PostgreSQL or MySQL database via your preferred ORM. During the setup of Acra, you will redirect your application to talk to Acra instead.
+- Use of TLS to access Acra and the database (our strong recommendation). You might need to configure a TLS certificate for Acra. Otherwise, you'll set up an encrypted transport layer using AcraConnector and [Secure Session]({{< ref "themis/crypto-theory/cryptosystems/secure-session.md" >}}).
+- [AcraWriter]({{< ref "acra/acra-in-depth/security-design/#acrawriter-INVALID" >}}) library installed for your application that will encrypt the data inside your app ("client side encryption"). AcraWriter uses [Themis]({{< ref "themis" >}}) and OpenSSL/BoringSSL as dependencies. If you prefer "transparent encryption proxy" mode, no additional libraries will be necessary for your app.
+- If you're installing Acra on your server manually from the [GitHub repository](https://github.com/cossacklabs/acra), you need to have Themis' dependencies and `libssl-dev` package installed. Also, make sure that `libcrypto.so` is available in `$PATH`.
```
I prefer to have hope in people's mental capabilities. If you're going to build from source, you're probably competent enough to find out where the dependencies are documented and resolve the issues.

It's also possible that people would like to build Acra from source while installing Themis from packages. If they are using packages, all dependencies are already handled there.

I believe that packaging Acra for more platforms would be more beneficial for usability than duplicating incomplete compilation instructions. Then people would not need to bother with this in the first place.

Lagovas commented on acra/guides/trying-acra-with-docker/_index.md:
```diff
@@ -0,0 +1,403 @@
+
+This is why this document is called "Trying Acra with Docker", not "Using Acra with Docker in a Production Setting".
+
+### Precautions
+
+Storing keys safely when using Docker is problematic. Docker is immutable while [Zones]({{< ref "acra/acra-in-depth/cryptography-and-key-management#zones-INVALID" >}})/[keys]({{< ref "acra/acra-in-depth/cryptography-and-key-management#-INVALID" >}}) are not. This means that you might want to attach some storage and end up making the keys accessible to attackers. There are multiple ways to solve t ...
```
Actually not. If it's a server where running only Acra then it's okay. But if it's some server where deployed different containers and you store keys on the host layer and mount it to Acra containers - it's not the same and not safe. Because this folder may be mounted to other containers. Or some container are vulnerable, can escalate permissions and get access to root FS. So keys are not safe in such cases that we can't ignore. And we can't check or handle it from Acra's containers.

Lagovas commented on acra/guides/trying-acra-with-docker/_index.md:
````diff
@@ -0,0 +1,403 @@
+
+All that you need for your first try of Acra is to launch the selected scheme from the `docker` subdirectory:
+
+```bash
+docker-compose -f docker/<compose_file_name>.yml up
+```    
````
agree

Lagovas commented on acra/guides/trying-acra-with-docker/_index.md:
```diff
@@ -0,0 +1,403 @@
+
+There are several different dockerfiles in `docker` subdirectory:
+
+* `acra-authmanager.dockerfile` – resulting image with AcraAuthmanager tool
+* `acra-connector.dockerfile` – resulting image with AcraConnector
+* `acra-server.dockerfile` – resulting image with AcraServer
+* `acra-keymaker.dockerfile` – resulting image with AcraKeymaker tool
+* `acra-webconfig.dockerfile` – resulting image with AcraWebconfig component
+* `acra-build.dockerfile` – intermediate image for compiling all the Acra components
+* `mysql-nossl.dockerfile` – MySQL server container with disabled SSL
+* `mysql-ssl.dockerfile` – MySQL server container with example SSL certificates (located at ssl/mysql directory)
+* `postgresql-ssl.dockerfile` – Postgresql server container with example SSL certificates (located at ssl/postgresql directory)
```
agree

Lagovas commented on acra/guides/trying-acra-with-docker/_index.md:
```diff
@@ -0,0 +1,403 @@
+
+All docker images have tags. When we build them, we set appropriate tags for each image:
+
+* sliding tags:
+    - `stable` and `latest` - stable branch, recommended, default;
+    - `master` and `current` - master branch of GitHub repository.
```
okay. I didn't meet any of them because release posts I read diagonally)

Lagovas commented:
Applied most of the comments where agree and it's critical. If something is missed (not all comments displayed here after some commits and not all comments displayed in `files changed` tab, so probably I may miss something. I will update/fix it in next PR's. Please mark as resolved comments where no more questions/suggestions.
This PR I will merge to master/staging and deploy to staging because it a bit blocks other processes.

@ilammy great thanks for review and comments) Didn't know before that `**1**` not rendered to `<ol>` *(
